File: iosApp/iosApp/Modules/QRCodeScan/QRCodeScanWireframe.swift
// Created by web3d4v on 21/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3lib

// MARK: - QRCodeScanWireframeContext

struct QRCodeScanWireframeContext {
    let type: `Type`
    let onCompletion: (String) -> Void

    enum `Type` {
        case `default`
        case network(Network)
    }
}

// MARK: - QRCodeScanWireframeDestination

enum QRCodeScanWireframeDestination {
    case qrCode(String)
}

// MARK: - QRCodeScanWireframe

protocol QRCodeScanWireframe {
    func present()
    func navigate(to destination: QRCodeScanWireframeDestination)
    func dismiss()
}

// MARK: - DefaultQRCodeScanWireframe

final class DefaultQRCodeScanWireframe {
    
    private weak var parent: UIViewController?
    private let context: QRCodeScanWireframeContext
    
    private weak var vc: UIViewController?
    
    init(
        _ parent: UIViewController?,
        context: QRCodeScanWireframeContext
    ) {
        self.parent = parent
        self.context = context
    }
}

extension DefaultQRCodeScanWireframe: QRCodeScanWireframe {
    
    func present() {
        let vc = wireUp()
        parent?.show(vc, sender: self)
    }
    
    func navigate(to destination: QRCodeScanWireframeDestination) {
        switch destination {
        case let .qrCode(qrCode):
            context.onCompletion(qrCode)
            dismiss()
        }
    }
    
    func dismiss() {
        vc?.popOrDismiss()
    }
}

private extension DefaultQRCodeScanWireframe {
    
    func wireUp() -> UIViewController {
        let interactor = DefaultQRCodeScanInteractor()
        let vc: QRCodeScanViewController = UIStoryboard(.qrCodeScan).instantiate()
        let presenter = DefaultQRCodeScanPresenter(
            view: vc,
            wireframe: self,
            interactor: interactor,
            context: context
        )
        vc.presenter = presenter
        vc.hidesBottomBarWhenPushed = true
        self.vc = vc
        return vc
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/qrcodescan/QRCodeScanWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.qrcodescan

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.R
import com.sonsofcrypto.web3wallet.android.common.NavigationFragment
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3walletcore.modules.nftSend.*
import com.sonsofcrypto.web3walletcore.modules.qrCodeScan.*

class DefaultQRCodeScanWireframe(
    private val parent: WeakRef<Fragment>?,
    private val context: QRCodeScanWireframeContext,
): QRCodeScanWireframe {

    private lateinit var fragment: WeakRef<Fragment>

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.present(fragment, animated = true)
    }

    override fun navigate(destination: QRCodeScanWireframeDestination) {
        when (destination) {
            is QRCodeScanWireframeDestination.QRCode -> {
                context.handler(destination.value)
            }
            is QRCodeScanWireframeDestination.Dismiss -> {
                parent?.navigationFragment?.popOrDismiss()
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = QRCodeScanFragment()
        val interactor = DefaultQRCodeScanInteractor()
        val presenter = DefaultQRCodeScanPresenter(
            WeakRef(view),
            this,
            interactor,
            context,
        )
        view.presenter = presenter
        return view
    }
}